File: backend/src/main/java/com/suplr/backend/controller/WhatsAppWebhookController.java
package com.suplr.backend.controller;

import com.suplr.backend.entity.Message;
import com.suplr.backend.repository.ClientRepository;
import com.suplr.backend.repository.SupplierRepository;
import com.suplr.backend.service.CacheService;
import com.suplr.backend.service.WhatsAppService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@Slf4j
@RestController
@RequestMapping("/webhook")
@RequiredArgsConstructor
public class WhatsAppWebhookController {

    private final WhatsAppService whatsAppService;
    private final CacheService cacheService;
    private final SupplierRepository supplierRepository;
    private final ClientRepository clientRepository;

    @Value("${app.webhook.verify-token:}")
    private String webhookVerifyToken;

    @GetMapping("/{supplierId}")
    public ResponseEntity<String> verify(
            @PathVariable Integer supplierId,
            @RequestParam(value = "hub.mode", defaultValue = "") String mode,
            @RequestParam(value = "hub.verify_token", defaultValue = "") String token,
            @RequestParam(value = "hub.challenge", defaultValue = "") String challenge
    ) {
        if ("subscribe".equals(mode) && webhookVerifyToken.equals(token)) {
            log.info("Webhook verified for supplier {}", supplierId);
            return ResponseEntity.ok()
                    .contentType(MediaType.TEXT_PLAIN)
                    .body(challenge);
        }
        log.warn("Webhook verification failed for supplier {} — bad token", supplierId);
        return ResponseEntity.status(403).body("Verification failed");
    }

    @PostMapping("/{supplierId}")
    public Map<String, String> receive(
            @PathVariable Integer supplierId,
            @RequestBody(required = false) Map<String, Object> jsonBody,
            HttpServletRequest request
    ) {
        String contentType = request.getContentType() != null
                ? request.getContentType() : "";

        if (contentType.contains("application/json") && jsonBody != null) {
            @SuppressWarnings("unchecked")
            List<Map<String, Object>> messages =
                    (List<Map<String, Object>>) jsonBody.getOrDefault("messages", List.of());

            for (Map<String, Object> msg : messages) {
                if (!"text".equals(msg.get("type"))) continue;
                String msgId = (String) msg.get("id");
                String from = (String) msg.get("from");
                @SuppressWarnings("unchecked")
                String body = (String) ((Map<String, Object>) msg.get("text")).get("body");
                if (from != null && body != null) {
                    process(supplierId, msgId, from, body);
                }
            }
        } else {
            String msgId = request.getParameter("MessageSid");
            String from = request.getParameter("From");
            String body = request.getParameter("Body");
            if (from != null) from = from.replace("whatsapp:", "");
            if (from != null && body != null && !body.isBlank()) {
                process(supplierId, msgId != null ? msgId : "", from, body);
            }
        }

        return Map.of("status", "ok");
    }

    private void process(Integer supplierId, String msgId, String fromNumber, String text) {
        try {
            // Idempotency check
            if (msgId != null && !msgId.isBlank() && cacheService.isMessageSeen(supplierId, msgId)) {
                log.info("Duplicate message {} for supplier {} — skipped", msgId, supplierId);
                return;
            }
            if (msgId != null && !msgId.isBlank()) {
                cacheService.markMessageSeen(supplierId, msgId);
            }

            Message message = whatsAppService.storeInboundMessage(
                    supplierId, msgId, fromNumber, text);

            if (whatsAppService.handleNameCollection(
                    supplierId, message.getClientId(), text, fromNumber)) {
                return;
            }

            if (whatsAppService.isHistoryQuery(text)) {
                String lang = clientRepository.findById(message.getClientId())
                        .map(c -> c.getPreferredLanguage() != null ? c.getPreferredLanguage() : "en")
                        .orElse("en");
                whatsAppService.handleHistoryQuery(
                        supplierId, message.getClientId(), fromNumber, lang);
                return;
            }

            if (whatsAppService.handlePendingConfirmation(
                    supplierId, message.getClientId(), fromNumber, text)) {
                return;
            }

            supplierRepository.findById(supplierId).ifPresent(supplier -> {
                if ("pro".equals(supplier.getPlan())) {
                    whatsAppService.parseAndCreateOrder(supplierId, message, fromNumber);
                }
            });

        } catch (Exception e) {
            log.error("Error processing message from {} for supplier {}: {}",
                    fromNumber, supplierId, e.getMessage(), e);
        }
    }
}

File: backend/src/whatsapp/whatsapp-webhook.controller.ts
import {
  Controller,
  Get,
  Post,
  Param,
  Query,
  Body,
  Req,
  ParseIntPipe,
  Logger,
} from '@nestjs/common';
import type { Request } from 'express';
import { WhatsAppService } from './whatsapp.service.js';
import { CacheService } from '../cache/cache.service.js';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Supplier } from '../entities/supplier.entity.js';
import { Client } from '../entities/client.entity.js';
import { ConfigService } from '@nestjs/config';

@Controller('webhook')
export class WhatsAppWebhookController {
  private readonly logger = new Logger(WhatsAppWebhookController.name);

  constructor(
    private readonly whatsAppService: WhatsAppService,
    private readonly cache: CacheService,
    @InjectRepository(Supplier)
    private readonly supplierRepo: Repository<Supplier>,
    @InjectRepository(Client) private readonly clientRepo: Repository<Client>,
    private readonly config: ConfigService,
  ) {}

  @Get(':supplierId')
  verify(
    @Param('supplierId', ParseIntPipe) supplierId: number,
    @Query('hub.mode') mode: string,
    @Query('hub.verify_token') token: string,
    @Query('hub.challenge') challenge: string,
  ) {
    const verifyToken = this.config.get<string>('WEBHOOK_VERIFY_TOKEN', '');
    if (mode === 'subscribe' && token === verifyToken) {
      this.logger.log(`Webhook verified for supplier ${supplierId}`);
      return challenge;
    }
    this.logger.warn(`Webhook verification failed for supplier ${supplierId}`);
    return 'Verification failed';
  }

  @Post(':supplierId')
  async receive(
    @Param('supplierId', ParseIntPipe) supplierId: number,
    @Body() jsonBody: Record<string, unknown>,
    @Req() req: Request,
  ) {
    const contentType = req.headers['content-type'] ?? '';

    if (contentType.includes('application/json') && jsonBody) {
      const messages = (jsonBody['messages'] as any[]) ?? [];
      for (const msg of messages) {
        if (msg.type !== 'text') continue;
        const msgId: string = msg.id ?? '';
        const from: string = msg.from ?? '';
        const body: string = (msg.text as any)?.body ?? '';
        if (from && body) await this.process(supplierId, msgId, from, body);
      }
    } else {
      const body = req.body as Record<string, string>;
      const msgId = body['MessageSid'] ?? '';
      let from = body['From'] ?? '';
      const text = body['Body'] ?? '';
      from = from.replace('whatsapp:', '');
      if (from && text) await this.process(supplierId, msgId, from, text);
    }

    return { status: 'ok' };
  }

  private async process(
    supplierId: number,
    msgId: string,
    fromNumber: string,
    text: string,
  ) {
    try {
      if (msgId && (await this.cache.isMessageSeen(supplierId, msgId))) {
        this.logger.log(
          `Duplicate message ${msgId} for supplier ${supplierId} — skipped`,
        );
        return;
      }
      if (msgId) await this.cache.markMessageSeen(supplierId, msgId);

      const message = await this.whatsAppService.storeInboundMessage(
        supplierId,
        msgId,
        fromNumber,
        text,
      );

      if (
        await this.whatsAppService.handleNameCollection(
          supplierId,
          message.clientId!,
          text,
          fromNumber,
        )
      )
        return;

      if (this.whatsAppService.isHistoryQuery(text)) {
        const client = await this.clientRepo.findOne({
          where: { id: message.clientId! },
        });
        const lang = client?.preferredLanguage ?? 'en';
        await this.whatsAppService.handleHistoryQuery(
          supplierId,
          message.clientId!,
          fromNumber,
          lang,
        );
        return;
      }

      if (
        await this.whatsAppService.handlePendingConfirmation(
          supplierId,
          message.clientId!,
          fromNumber,
          text,
        )
      )
        return;

      const supplier = await this.supplierRepo.findOne({
        where: { id: supplierId },
      });
      if (supplier?.plan === 'pro') {
        await this.whatsAppService.parseAndCreateOrder(
          supplierId,
          message,
          fromNumber,
        );
      }
    } catch (e: any) {
      this.logger.error(
        `Error processing message from ${fromNumber} for supplier ${supplierId}: ${e.message}`,
        e.stack,
      );
    }
  }
}
